src/java.base/share/classes/jdk/internal/util/ArraysSupport.java

```diff
@@ -164,17 +166,177 @@ public static int vectorizedMismatch(Object a, long aOffset,
+            case 1 -> 31 * initialValue + (int) a[fromIndex];
+            default -> vectorizedHashCode(a, fromIndex, length, initialValue, T_CHAR);
+        };
+    }
+
+    /**
+     * Calculates the hash code for the subrange of a byte array.
+     *
+     * <p> This method does not perform type checks or bounds checks. It is the
+     * responsibility of the caller to perform such checks before calling this
+     * method.
+     *
+     * @param a the array
+     * @param fromIndex the first index of the subrange of the array
+     * @param length the number of elements in the subrange
+     * @param initialValue the initial hash value, typically 0 or 1
+     *
+     * @return the calculated hash value
+     */
+    public static int hashCode(byte[] a, int fromIndex, int length, int initialValue) {
+        return switch (length) {
+            case 0 -> initialValue;
+            case 1 -> 31 * initialValue + (int) a[fromIndex];
+            default -> vectorizedHashCode(a, fromIndex, length, initialValue, T_BYTE);
+        };
+    }
+
+    /**
+     * Calculates the hash code for the subrange of a byte array whose elements
+     * are treated as unsigned bytes.
+     *
+     * <p> This method does not perform type checks or bounds checks. It is the
+     * responsibility of the caller to perform such checks before calling this
+     * method.
+     *
+     * @param a the array
+     * @param fromIndex the first index of the subrange of the array
+     * @param length the number of elements in the subrange
+     * @param initialValue the initial hash value, typically 0 or 1
+     *
+     * @return the calculated hash value
+     */
+    public static int hashCodeOfUnsigned(byte[] a, int fromIndex, int length, int initialValue) {
+        return switch (length) {
+            case 0 -> initialValue;
+            case 1 -> 31 * initialValue + (a[fromIndex] & 0xff);
+            default -> vectorizedHashCode(a, fromIndex, length, initialValue, T_BOOLEAN);
+        };
+    }
+
+    /**
+     * Calculates the hash code for the subrange of a byte array whose contents
+     * are treated as UTF-16 chars.
+     *
+     * <p> This method does not perform type checks or bounds checks. It is the
+     * responsibility of the caller to perform such checks before calling this
+     * method.
+     *
+     * <p> {@code fromIndex} and {@code length} must be scaled down to char
+     * indexes.
+     *
+     * @param a the array
+     * @param fromIndex the first index of a char in the subrange of the array
+     * @param length the number of chars in the subrange
+     * @param initialValue the initial hash value, typically 0 or 1
+     *
+     * @return the calculated hash value
+     */
+    public static int hashCodeOfUTF16(byte[] a, int fromIndex, int length, int initialValue) {
+        return switch (length) {
+            case 0 -> initialValue;
+            case 1 -> 31 * initialValue + JLA.getUTF16Char(a, fromIndex);
```

Comment:
There seems to be a mismatch here with the original code in StringUTF16, where the length that is tested for is `2` instead of `1`.

Reply:
Yes, should be `2` (`a` is semantically a `char[]`). This typo likely pass functional testing since `1` can never happen in practice, and the default case should work for any value. There might be a String microbenchmark out there that might be slightly unhappy, though.

Reply:
I believe, it should be `1`. Hear me out. In this method, the `length` is scaled down, whereas in `StringUTF16` it is not. In this method, it's `length`, in StringUTF16 it's `((byte[]) value).length`.

Reply:
In fact, if I change it to `2`, the following tests will fail:

  - `jdk/jdk/classfile/Utf8EntryTest.java`
  - `jdk/java/util/zip/ZipCoding.java`
  - `jdk/java/text/Format/MessageFormat/MessageRegression.java`

Reply:
Indeed, the actual length passed at call site is `value.length >> 1` instead of `value.length`; this adjusted char-length carries on to `vectorizedHashCode` call.

Reply:
Ah, sneaky. Might affect scores for the zero and one-char cases since the shift now happens unconditionally, but probably doesn't matter in practice.